0005-configuration-management-extension/README.md

```diff
@@ -3,102 +3,151 @@ name: Configuration Management Extension
 
-2. _The server should know if a configuration has been delivered._ 
+## Requirements and constraints
 
-   Solutions:
-   - Use request/response messaging pattern from [Kaa Protocol RFC1](/0001-kaa-protocol/README.md).
+- CMX protocol must support client-initiated configuration data pull.
+Compliant endpoints should be able to receive the latest configuration by request.
+- In case of the configuration pull no receipt confirmation is required from the client.
+- CMX protocol must support server-initiated configuration data push.
+Compliant endpoints should be able to receive the latest configuration data whenever it changes at the server or at the moment of connecting without explicitly requesting a configuration update.
+- In case of the configuration push compliant clients MUST acknowledge the receipt.
+- In cases when the configuration data changes more than once while the endpoint is not connected to the server, after reconnecting it SHOULD receive only the latest data.
+- CMX protocol implementations MUST support JSON-structured configuration data transfer.
+Support for any other formats is OPTIONAL.
 
-## Use cases
+## Design
 
-### UC1: pull
-Configuration delivery by request. The endpoint should be able to receive latest configuration from CMX by request.
+### Configuration pull
 
-### UC2: push
-Configuration delivery that is initiated by the server. The endpoint should receive latest configuration when it connects to a server if this configuration hadn't applied yet.
+#### Configuration pull request
 
-## Design
+Client MUST send requests with the following extension-specific [resource path](/0001-kaa-protocol/README.md#resource-path-format) part in order to receive the configuration:
 
-### Request/response
-This topic is covered in [Kaa Protocol RFC1](/0001-kaa-protocol/README.md)
+  `<endpoint_token>/pull/<message_format>[/<config_format>]`
 
+  where:
+  - `endpoint_token` identifies the endpoint;
+  - `message_format` specifies the format of the configuration pull messages' payload, e.g. `json`, `protobuf`, etc.;
+  - `config_format` specifies the format of the configuration body, e.g. `json`, `avro`, etc.
+  This resource path parameter is optional.
+  In case it is not specified, the configuration body format is assumed to match the message format.
+  It is RECOMMENDED to omit the `<config_format>` resource path suffix in case the desired configuration format matches the message format.
 
-### Formats
-#### Schemeless JSON
-##### Configuration pull
-Endpoint must send messages to the next topic in order to obtain configuration:
-```
-<endpoint_token>/pull/json
-```
+Example configuration pull request extension-specific resource path parts:
+  - `<endpoint_token>/pull/json`
+  - `<endpoint_token>/pull/json/json`
+  - `<endpoint_token>/pull/json/avro`
```

Comment:
Specific Avro scheme will already be defined by the appversion in the common resource path. No need to duplicate it, I think.